core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,220 @@
+                    // Proto scalar types
+                case "float":
+                case "double":
+                    return "number";
+                case "i":
+                case "i32":
+                case "i64":
+                case "integer":
+                case "int":
+                case "l32":
+                case "l64":
+                case "long":
+                case "long32":
+                case "long64":
+                    // Proto scalar types
+                case "int32":
+                case "int64":
+                case "uint32":
+                case "uint64":
+                case "sint32":
+                case "sint64":
+                case "fixed32":
+                case "fixed64":
+                case "sfixed32":
+                case "sfixed64":
+                    return "integer";
+                case "binary":
+                    // Proto scalar types
+                case "bytes":
+                case "string":
+                    return "string";
+                default:
+                    return "null";
+            }
+        }
+
+        return "any";
+    }
+
+    /**
+     * Generate a FieldSchemaWithAdditionalProperties for the given fields.
+     * @param fields list of fields that the child has.
+     * @param visited a map of visited fields, required for cycle detection.
+     * @param path current path as defined in JSON Schema spec, required for cyclic references.
+     * @return a FieldSchemaWithAdditionalProperties for the given fields.
+     */
+    private static FieldSchemaWithAdditionalProperties generateFields(List<FieldInfo> fields,
+                                                                      Map<String, String> visited,
+                                                                      String path) {
+        final ObjectNode objectNode = mapper.createObjectNode();
+        final Set<String> requiredFields = new HashSet<>();
+
+        for (FieldInfo field : fields) {
+            final ObjectNode fieldNode = mapper.createObjectNode();
+            final String fieldTypeName = field.typeSignature().name();
+
+            if (field.typeSignature().isNamed() && visited.containsKey(fieldTypeName)) {
+                // If field is already visited, add a reference to the field instead of iterating its children.
+                final String pathName = visited.get(fieldTypeName);
+                fieldNode.put("$ref", pathName);
+            } else {
+                // Field is not visited, create a new type definition for it.
+                final String schemaType = getSchemaType(field.typeSignature());
+                fieldNode.put("type", schemaType);
+                fieldNode.put("description", field.descriptionInfo().docString());
+
+                // TODO: Support for repeated fields with non-primitive types.
+                // Use "items": { ... }
+                // But unfortunately container types do not contain field infos.
```

Comment:
Can't we use the `name()`?
```java
TypeSignature type = ...
if (type.isContainer() && ("list".equals(type.name()) || "set".equals(type.name()))) {
    TypeSignature parameterType = type.typeParameters().get(0);
}
```

Reply:
This is mostly fine, I am struggling with writing some re-usable code here and there but this should do the job for at least the first iteration. 

Trying to perfect those container types + enums is going to be very tough and we can even consider writing that logic platform specific. I have faced some issues with annotated services for example, I couldn't find a way to determine whether a `type` or `field` is enum or not (unlike grpc which can be inferred from its descriptor).

Reply:
So I suggest adding some tech-debt tickets for future iterations. Drilling those edge cases one by one can be easier and will probably result in much cleaner code.

Reply:
That is correct. 😄 
Let me send a PR that hides all these implementation details. 😄